### PR TITLE
Enable prune preserve quantization aware training (pqat)

### PR DESCRIPTION
**Pull request by psunn**
This PR intend to introduce Prune preserve Quant Aware Training

Changes:    
* Implemented pqat registry for supported keras layers.
* Implemented pqat weights quantizers.
* Example with simple cnn model on mnist.
* Added unit tests around the new changes.
* Public API `tfmot.experimental.combine.Default8BitPrunePreserveQuantizeScheme()`

It works as: 
```
import tensorflow_model_optimization as tfmot

(...)

tfmot.quantization.keras.quantize_apply(quant_aware_annotate_model,
    scheme=tfmot.experimental.combine.Default8BitPrunePreserveQuantizeScheme())
```
(please check _mnist_cnn.py_ as example)

Experiment results:

|Model | Base | Prune | QAT | PQAT |
| -- | -- | -- | -- | --|
|DS-CNN-L | <ul><li>FP32 Top 1 Acc: **95.233**%</li></ul> <ul><li>FP32 .tflite compressed size: (**1651512** -> **1530470**)</li></ul> <ul><li>INT8 Top 1 Acc: **94.476**%</li></ul> <ul><li>INT8 .tflite compressed size: (**528128** -> **434879**)</li></ul> | <ul><li>FP32 Top 1 Acc: **94.803**%</li></ul> <ul><li>FP32 .tflite compressed size: (**1651512** -> **946584**)</li></ul> <ul><li>INT8 Top 1 Acc: **93.801**%</li></ul> <ul><li>INT8 .tflite compressed size: (**528128** -> **334154**)</li></ul> <ul><li>model sparsity: **50**%</li></ul> | <ul><li>INT8 Top 1 Acc: **94.721**%</li></ul> <ul><li>INT8 .tflite compressed size: (**512224** -> **403261**)</li></ul> | <ul><li>INT8 Top 1 Acc: **94.128**%</li></ul> <ul><li>INT8 .tflite compressed size: (**512032** -> **303997**)</li></ul> <ul><li>model sparsity: **50**%</li></ul>|
|MobileNet-V1-224 | <ul><li>FP32 Top 1 Acc: **70.99**%</li></ul> <ul><li>FP32 .tflite compressed size:(**16908804** → **15682762**)</li></ul> <ul><li>INT8 Top 1 Acc: **69.37**%</li></ul> <ul><li>INT8 .tflite compressed size:(**4665520** → **3880331**)</li></ul> | <ul><li>FP32 Top 1 Acc: **70.11**%</li></ul> <ul><li>FP32 .tflite compressed size: (**16,908,804** → **9,553,824**)</li></ul> <ul><li>INT8 Top 1 Acc: **67.82**%</li></ul> <ul><li>INT8 .tflite compressed size: (**4,665,520** → **2,939,734**)</li></ul> <ul><li>model sparsity: **50**%</li></ul> | <ul><li>INT8 Top 1 Acc: **70.67**%</li></ul> <ul><li>INT8 .tflite compressed size: (**4,569,416** → **3,808,781**)</li></ul> | <ul><li>INT8 Top 1 Acc: **70.29**%</li></ul> <ul><li>INT8 .tflite compressed size:(**4,569,416** → **2,869,600**)</li></ul> <ul><li>model sparsity: **50**%</li></ul>|

> model size in bytes

### CODE REVIEW

**Comment by psunn** [comment]
@alanchiao @nutsiepully please have a review.

File: tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/prune_preserve_quantize_registry.py [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+      # SeparableConv need verify from 8bit qat
+      # layers.SeparableConv1D:
+      # _PrunePreserveInfo(['pointwise_kernel'], ['Default8BitConvQuantizeConfig']),
+      # layers.SeparableConv2D:
+      # _PrunePreserveInfo(['pointwise_kernel'], ['Default8BitConvQuantizeConfig']),
+
+      # Embedding need verify from 8bit qat
+      # layers.Embedding: _PrunePreserveInfo(['embeddings'], []),
+
+      layers.ELU: _noPrunePreserve(),
```
Comment (alanchiao) [inline tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/prune_preserve_quantize_registry.py:87]: You can use the approach used by Elena under https://github.com/tensorflow/model-optimization/pull/539 to cut almost all of the _noPrunePreserve lines.

File: tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/prune_preserve_quantize_registry.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+      layers.GlobalMaxPooling1D: _noPrunePreserve(),
+      layers.GlobalMaxPooling2D: _noPrunePreserve(),
+      layers.GlobalMaxPooling3D: _noPrunePreserve(),
+      layers.MaxPooling1D: _noPrunePreserve(),
+      layers.MaxPooling2D: _noPrunePreserve(),
+      layers.MaxPooling3D: _noPrunePreserve(),
+      TensorFlowOpLayer: _noPrunePreserve(),
+  }
+
+  # TODO(mltools): enable RNN layers, also depend on qat and prune registry
```
Comment (alanchiao) [inline tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/prune_preserve_quantize_registry.py:141]: Depending on the registries directly needs to be discussed since at the moment, those are private APIs.

File: tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/prune_preserve_quantize_registry_test.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import numpy as np
+import tensorflow as tf
+
+from tensorflow.python.keras import keras_parameterized
+
+from tensorflow_model_optimization.python.core.quantization.keras import quantize_config
+from tensorflow_model_optimization.python.core.sparsity.keras import prune_registry
+from tensorflow_model_optimization.python.core.quantization.keras.default_8bit import default_8bit_quantize_registry
+from tensorflow_model_optimization.python.core.quantization.keras.prune_preserve import prune_preserve_quantize_registry
+
```
Comment (alanchiao) [inline tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/prune_preserve_quantize_registry_test.py:28]: When there is further agreement, there should be tests that actually verify that the sparsity is preserved after some quantization-aware training.

File: tensorflow_model_optimization/python/examples/quantization_with_sparsity/keras/mnist_cnn.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
```
Comment (alanchiao) [inline tensorflow_model_optimization/python/examples/quantization_with_sparsity/keras/mnist_cnn.py:1]: It'd be good to consider the discussion in https://github.com/tensorflow/model-optimization/pull/562 together with this.

File: tensorflow_model_optimization/python/core/quantization/keras/quantize.py [hunk 5]
```diff
@@ -361,6 +376,11 @@ def _quantize(layer):  # pylint: disable=missing-docstring
     if not quantize_config and quantize_registry.supports(layer):
       quantize_config = quantize_registry.get_quantize_config(layer)
 
+    # (experimental) enable sparsity preserved quantization
```
Comment (alanchiao) [inline tensorflow_model_optimization/python/core/quantization/keras/quantize.py:379]: To discuss regarding the API with Pulkit. I'm inclined to keep this experimental functionality separate from the existing stable quantization-aware-training APIs, until there has been an actual demonstrated value for it (e.g. some model where pruning + sparsity-preserving quantization-aware training does significantly better than pruning + post-training quantization). 

This logic can be refactored to minimize code duplication while preserving readability of the overall flow.

Some of the issues (BN + Concat) in https://github.com/tensorflow/model-optimization/issues/513#issuecomment-673603923 haven't been addressed either, for getting something checked into this stable code.

File: tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/prune_preserve_quantize_registry.py (hunk 1 above)
Comment (psunn) [reply tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/prune_preserve_quantize_registry.py]: thanks!
done, please have a review.

File: tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/prune_preserve_quantize_registry.py (hunk 2 above)
Comment (psunn) [reply tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/prune_preserve_quantize_registry.py]: removed

File: tensorflow_model_optimization/python/core/quantization/keras/quantize.py (hunk 5 above)
Comment (psunn) [reply tensorflow_model_optimization/python/core/quantization/keras/quantize.py:379]: Regarding (BN + Concat):  current implementation, pqat only update weights quantizers after quantize transform, thus cross-layer modifications remains.

File: tensorflow_model_optimization/python/examples/quantization_with_sparsity/keras/mnist_cnn.py (hunk 4 above)
Comment (psunn) [reply tensorflow_model_optimization/python/examples/quantization_with_sparsity/keras/mnist_cnn.py:1]: Thanks for point it out. Did we have conclusion regarding example scripts?

**Comment by nutsiepully** [comment]
Hi @psunn,

Thanks for the work on this. I have a few broad comments on the overall direction of this CL.

For starters, we shouldn't need to modify any QAT code under `python/core/quantization/keras` to achieve this. New quantization schemes should just be plugged in. The quantization code cannot be aware of specific flags such as `prune_preserve`.

We can add a new parameter `quantization_scheme` which was planned to be added to the `quantize` API which you can use to plug in your scheme. So you should only need to implement a `PrunePreserveQuantizationScheme`. Not much else.

Also, I don't think we need a new `prune_preserve` directory in `model-optimization` just yet. I think you should be able to depend on MOT library, write your own scheme and try it out.

If the results are good, and we want to include it, we can then add it as a scheme under quantization. So, just like `default_8bit_scheme`, we can have a folder for `default_8bit_pruned_scheme` or something similar.

Does that seem reasonable? Thanks a lot for your work on this.

cc @alanchiao

**Comment by psunn** [comment]
Hi @nutsiepully,

Thanks for review.

Prune preserve is for prunable weights, so PQAT is to modify the Ops in weights quantizers. In my opinion, prune preserve quantization shouldn't be another stand alone quantization configuration.

In current TFMOT quantization, `quantize_config` defines how weights, activations, outputs should be quantized with corresponding `Quantizer`. In the current proposal, PQAT is applied after `QuantizeLayoutTransform`, by replacing `weight_quantizer` with `PrunePreserveQuantizer` for registered layers with registered `quantize_config`.

I'm simply guessing the upcoming `quantization_scheme` is holding both `QuantizeLayoutTransform` and `quantize_config`, nice for customizations. Prune preservation is a general feature for QAT, it's nice that each `quantization_scheme` is able to switch it on/off. So maybe it's unnecessary for prune preserve QAT to stand as a separate `quantization_scheme` as well? I'm not sure how `quantization_scheme` gonna looks like, how about prune preserve QAT to align and refactoring after `quantization_scheme` is available in api?

**Comment by nutsiepully** [comment]
This is what the code would like.

```
class PrunePreseveQuantizationScheme(QuantizationScheme):
 def get_model_transformer(self):
   return Default8BitModelTransformer()

 def get_quantize_registry(self):
   return PrunePreserveDefault8BitQuantizeRegistry()

quantized_model = tfmot.quantization.keras.quantize_model(
   model, scheme=PrunePreseveQuantizationScheme())
```

**Comment by psunn** [comment]
@alanchiao @nutsiepully thanks for https://github.com/tensorflow/model-optimization/pull/591
I've update PQAT through QuantizationScheme, can you please have a review?

It works as:
```
import tensorflow_model_optimization as tfmot
import tensorflow_model_optimization.quantization.keras.experimental_scheme as experimental_scheme

(...)

tfmot.quantization.keras.quantize_apply(quant_aware_annotate_model,
    scheme=experimental_scheme.Default8BitPrunePreserveQuantizeScheme())
```

File: tensorflow_model_optimization/python/core/quantization/keras/quantize.py (hunk 5 above)
Comment (alanchiao) [reply tensorflow_model_optimization/python/core/quantization/keras/quantize.py]: > Regarding (BN + Concat): current implementation, pqat only update weights quantizers after quantize transform, thus cross-layer modifications remains. Am I missing something?

No you're not missing something. It's a non-issue.

File: tensorflow_model_optimization/python/examples/quantization_with_sparsity/keras/mnist_cnn.py (hunk 4 above)
Comment (alanchiao) [reply tensorflow_model_optimization/python/examples/quantization_with_sparsity/keras/mnist_cnn.py:1]: I think it's appropriate to not have the scripts in this case and substitute them with e2e unit tests (that don't take too long to run) and other documentation.

File: tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/BUILD [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    srcs = [
+        "__init__.py",
+    ],
+    srcs_version = "PY3",
+    deps = [
+        ":default_8bit_prune_preserve_quantize_scheme",
+    ],
+)
+
+py_library(
```
Comment (alanchiao) [inline tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/BUILD:18]: Would make the registry and any libraries that don't need to be used externally private in terms of visibility.

File: tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/prune_preserve_quantize_registry.py [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+
+  _DISABLE_PRUNE_PRESERVE = {
+      layers.DepthwiseConv2D,
+  }
+
+  def __init__(self):
+
+    self._config_quantizer_map = {
+        'Default8BitQuantizeConfig':
+        PrunePerserveDefault8BitWeightsQuantizer(),
```
Comment (alanchiao) [inline tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/prune_preserve_quantize_registry.py:93]: "PrunePerserve" -> PrunePreserve, other cases also

Comment (psunn) [reply tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/prune_preserve_quantize_registry.py]: Thanks!
Done

File: tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/BUILD (hunk 6 above)
Comment (psunn) [reply tensorflow_model_optimization/python/core/quantization/keras/prune_preserve/BUILD:18]: Done

**Comment by psunn** [comment]
@alanchiao @nutsiepully please have a review

Moved prune_preserve source folder  into collaborative_optimizations.
Changed public API as `tfmot.quantization.keras.collaborative_optimization_scheme.Default8BitPrunePreserveQuantizeScheme`
Further experiments with MobileNet_v1_224 shows better results (see top of the page).

**Comment by alanchiao** [comment]
Similar to https://github.com/tensorflow/model-optimization/pull/520:

Looks good to me as long as it's clear that it's an experimental API (e.g. tfmot.quantization.keras.experimental.Default8BitPrunePreserveQuantizeSchemes). Now that you have good results in the experiments, the most important thing right now is to gauge community interest in this feature by explicitly soliciting for it in an issue and also seeing if any models produced become useful enough on the accuracy / size curve that they'd practically be used - if it isn't large enough, we can remove it given the experimental nature. If it is of interest, we can finalize how it's exposed.

In other threads, Pulkit Bhuwalka had suggested exposing features like this not as a part of the individual clustering / quantization APIs, but rather as a module targeted towards different forms of combining techniques.

In this case, since the QuantizationScheme class is clearly there as an input to the quantize_apply API, I think exposing it in some common module like tfmot.combine (Pulkit had the idea tfmot.optimize) would be fine. The modules of the form "tfmot.[technique].keras" are focused on using one training-time technique at a time, which as far as we know so far, is the most common pattern right now and the common module has extra utilities to combine them.

**Comment by alanchiao** [comment]
@psunn: don't need to do the change right now but tfmot.quantization.keras.collaborative_optimization_scheme.Default8BitPrunePreserveQuantizeScheme isn't good since it's not clear from the symbol that it's experimental. 

Something like tfmot.quantization.keras.experimental.Default8BitPrunePreserveQuantizeScheme or tfmot.experimental.combine.Default8BitPrunePreserveQuantizeScheme would be better.

**Comment by psunn** [comment]
@alanchiao following our discussion, I've update public API as `tfmot.experimental.combine.Default8BitPrunePreserveQuantizeScheme()`.
Please have a review.

**Comment by psunn** [comment]
@alanchiao @nutsiepully could we move on to merge this PR now? what do you think?

**Comment by psunn** [comment]
@daverim @alanchiao @nutsiepully
Can someone help to review this PR, hopefully merge it if it's ok.

Hi david, it was suggested to contact you for TFMOT, can you help to proceed?

**Review by daverim — APPROVED** [review]
This looks fine to me
